cloud/scope/vpc_cluster.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scope
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/go-logr/logr"
+
+	"github.com/IBM/go-sdk-core/v5/core"
+	"github.com/IBM/platform-services-go-sdk/resourcecontrollerv2"
+
+	"k8s.io/klog/v2/textlogger"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	capiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/patch"
+
+	infrav1beta2 "sigs.k8s.io/cluster-api-provider-ibmcloud/api/v1beta2"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/services/authenticator"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/services/cos"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/services/resourcecontroller"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/services/resourcemanager"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/services/vpc"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/endpoints"
+)
+
+const (
+	// LOGDEBUGLEVEL indicates the debug level of the logs.
+	LOGDEBUGLEVEL = 5
+)
+
+// VPCClusterScopeParams defines the input parameters used to create a new VPCClusterScope.
+type VPCClusterScopeParams struct {
+	Client          client.Client
+	Cluster         *capiv1beta1.Cluster
+	IBMVPCCluster   *infrav1beta2.IBMVPCCluster
+	Logger          logr.Logger
+	ServiceEndpoint []endpoints.ServiceEndpoint
+
+	IBMVPCClient vpc.Vpc
+}
+
+// VPCClusterScope defines a scope defined around a VPC Cluster.
+type VPCClusterScope struct {
+	logr.Logger
+	Client      client.Client
+	patchHelper *patch.Helper
+
+	COSClient                cos.Cos
+	ResourceControllerClient resourcecontroller.ResourceController
+	ResourceManagerClient    resourcemanager.ResourceManager
+	VPCClient                vpc.Vpc
+
+	Cluster         *capiv1beta1.Cluster
+	IBMVPCCluster   *infrav1beta2.IBMVPCCluster
+	ServiceEndpoint []endpoints.ServiceEndpoint
+}
+
+// NewVPCClusterScope creates a new VPCClusterScope from the supplied parameters.
+func NewVPCClusterScope(params VPCClusterScopeParams) (*VPCClusterScope, error) {
```

Comment:
I am still not clear on how this will be invoked and embedded into existing flow, Do you mind adding that snippet of code to this PR. Also if possible we can move api definition to seperate PR if possible.

Reply:
It gets utilized here, in this new v2 path, which allows me to modify the `Cluster` resource without affecting the existing VPC (legacy) support:
https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1858/files#diff-647a58f1c11f771d3ec32c60c28c4b8b29442f9915129e16eeddb59120dcc7d8R140-R146

I can attempt to break up this PR, and maybe see if adding anything more for the new VPCClusterScope use that doesn't balloon the PR too much.

Reply:
>It gets utilized here, in this new v2 path, which allows me to modify the Cluster resource without affecting the existing VPC (legacy) support:
I got that but where do you call reconcilerv2?, Something like you need to add a check to see whether vpcclusterspec.NetworkSpec is set or not, If not set go with v1 flow and if set go with new v2 flow?


If this is the case, now if you are expecting user to go with v2 flow say for IPI case but user misses to set networkspec now the controller by default with v1 flow which is not you are expecting and we cannot make networkspec mandatory also, So we dont have any way to add a validation for v2 flow. Its up-to user intelligence and documentation to go with v2 flow..

Reply:
Yes, that is my expectation, the burden is on the user to provide details in order to utilize the v2 flow.
The architecture deployment for the v2 flow is going to be drastically different than the existing deployment.
In order to achieve and configure this, the user will be on the hook to configure the Cluster resource properly.

Reply:
>I got that but where do you call reconcilerv2?, Something like you need to add a check to see whether vpcclusterspec.NetworkSpec is set or not, If not set go with v1 flow and if set go with new v2 flow?

That is right here
https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1858/files#diff-647a58f1c11f771d3ec32c60c28c4b8b29442f9915129e16eeddb59120dcc7d8R74-R77

Otherwise, it continues with the legacy (v1) flow

Reply:
got it thanks